### PR TITLE
ARIES-1219 Weaving of Aries SPI Fly bundle produces RuntimeException: JSR/RET are not supported with computeFrames option

### PR DESCRIPTION
**Pull request by stiemannkj1**
Fixes https://issues.apache.org/jira/browse/ARIES-1219

### CODE REVIEW

**Comment by bosschaert** [comment]
It's been a while since I looked at this code, but on the surface the change looks good to me.

**Comment by vbansal22** [comment]
Can we merge this please? I have run into to a similar issue. Would like to see this fixed. I have reviewed the code and it looks good to me.

**Comment by bosschaert** [comment]
@timothyjward would you have time to review this?

**Comment by vbansal22** [comment]
Also tested the fix. Thanks @stiemannkj1 
It works well. Please merge.

**Comment by stiemannkj1** [comment]
@vbansal22 unfortunately, I don't have the ability to merge this. Still awaiting review.

**Comment by timothyjward** [comment]
> @timothyjward would you have time to review this?

This code seems overly complex… I’m not sure why the Java 8 check is present? Is there a reason not to include JSR inlining at all times?

File: spi-fly/spi-fly-itests/test.bndrun
Comment (stiemannkj1) [inline spi-fly/spi-fly-itests/test.bndrun]: These changes took place automatically when I performed a local build, so I think they might have been missed in a previous update.

**Comment by stiemannkj1** [comment]
> I’m not sure why the Java 8 check is present? Is there a reason not to include JSR inlining at all times?

Only performance. I can add a comment to that effect but Java 8+ compiled classes are guaranteed not have JSR/RET instructions. I believe weaving takes place for every class loaded after SPI-Fly is activated, so it could be quite costly to apply these unnecessary checks when the vast majority of classes are Java 8+ these days. I'd like to keep the check along with a comment as I don't think it adds significant complexity.

**Comment by wilx** [comment]
Not everyone using this is running Java 8+ code. I have some legacy apps that use OSGi that use SPI Fly where I had this issue because they are using components compiled as Java 5 (or older). They are using [Doug Lea's concurrency](http://gee.cs.oswego.edu/dl/). Rewriting the code to use modern concurrency classes is not an option, unfortunately. Beside some Java 5 (or older), many of the dependencies are Java 6 or 7, too.

I appreciate that my case is not typical. But if it can be supported with little to no penalty, that would be great.

**Comment by timothyjward** [comment]
> I appreciate that my case is not typical. But if it can be supported with little to no penalty, that would be great.

@wilx I don't think that there is any need to drop support for older bytecode versions, and there's nothing in this PR that would do so.

> Only performance. I can add a comment to that effect but Java 8+ compiled classes are guaranteed not have JSR/RET instructions.

@stiemannkj1 - that's what I thought. Can we remove those parts of this PR please? It's not good practice to mix bug fixes with feature enhancements, and it will confuse any regression analysis later if there are problems. I believe that this is also the wrong point in the code to optimise performance.

>  I believe weaving takes place for every class loaded after SPI-Fly is activated, so it could be quite costly to apply these unnecessary checks when the vast majority of classes are Java 8+ these days.

SPI-Fly already ignores bundles that haven't opted in to weaving, which is most of them, so this isn't as big a deal as you might think. If we were looking to improve performance further then we would:

1. Need to do some measurements. If we don't do this then we have no idea if it's worth it.
2. Look at higher-level options for skipping work, e.g. a two-stage check/weave model.

The current implementation writes a woven class file every time, but throws it away if there were no calls to decorate. Given that most class files don't need decoration (even in an opted-in bundle) it would be much faster to run a "check" which ignores everything but the method bodies to determine whether decoration is needed. Only if decoration is needed do we then run the full weaving code and generate the updated class bytes. The checker could also identify:

* Whether any methods contain JSR instructions so that decorator is only added if needed.
* Which methods need decoration so that we can avoid unnecessary re-checking in the weaving code.

**Comment by raghavendrav** [comment]
Hi @stiemannkj1 ,
The below file is missing a header and is getting flagged when doing a mvn clean install on spi-fly.
I worked around by adding the header myself locally.
[TCCLSetterVisitorTest.java]

**Review by timothyjward — CHANGES_REQUESTED** [review]
This commit makes too many unrelated changes. It should be cleaned up to simply inline JSR instructions, and if performance is an issue then a separate performance bug should be raised, along with measurements to show that the additional complexity is actually worthwhile.

**Comment by stiemannkj1** [comment]
@timothyjward, simplified to add the bare minimum of `JSRInlinerAdapter`. Should be good to merge now unless you notice any other issues.

**Review by timothyjward — APPROVED** [review]

**Comment by stiemannkj1** [comment]
@timothyjward can you merge when you get the chance? I don’t have permission. Thanks!

**Comment by vbansal22** [comment]
Thanks @stiemannkj1  and @timothyjward

**Comment by vbansal22** [comment]
Hi @timothyjward , is it possible to trigger a release build (1.3.7) after the merge?

**Pull request merged**
